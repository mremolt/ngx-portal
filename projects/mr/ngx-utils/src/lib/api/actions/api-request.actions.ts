import { HttpErrorResponse, HttpRequest } from '@angular/common/http';
import { Action } from '@ngrx/store';

import { ApiActions, ApiRequestConfig } from '../types';

export enum ApiRequestActionTypes {
  ApiRequest = '[API] Request',
  ApiError = '[API] Error',
}

export class ApiRequest<T = any> implements Action {
  public readonly type = ApiRequestActionTypes.ApiRequest;

  constructor(public payload: ApiRequestConfig<T>) {}
}

export class ApiError implements Action {
  public readonly type = ApiRequestActionTypes.ApiError;

  constructor(public payload: HttpErrorResponse) {}
}

export class ApiGetAction<T> extends ApiRequest {
  constructor(url: string, actions: ApiActions<T>) {
    const request = new HttpRequest<T>('GET', url);
    super({ request, actions });
  }
}

export class ApiPostAction<T> extends ApiRequest {
  constructor(url: string, body: T | null, actions: ApiActions<T>) {
    const request = new HttpRequest<T>('POST', url, body);
    super({ request, actions });
  }
}

export class ApiPutAction<T> extends ApiRequest {
  constructor(url: string, body: T | null, actions: ApiActions<T>) {
    const request = new HttpRequest<T>('PUT', url, body);
    super({ request, actions });
  }
}

export class ApiDeleteAction<T> extends ApiRequest {
  constructor(url: string, actions: ApiActions<T>) {
    const request = new HttpRequest<T>('DELETE', url);
    super({ request, actions });
  }
}
