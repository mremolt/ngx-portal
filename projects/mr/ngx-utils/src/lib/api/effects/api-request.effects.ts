import { HttpClient, HttpRequest, HttpResponse } from '@angular/common/http';
import { Inject, Injectable } from '@angular/core';
import { Actions, Effect, ofType } from '@ngrx/effects';
import { Store, select } from '@ngrx/store';
import { EMPTY, concat, of } from 'rxjs';
import { catchError, filter, flatMap, map, share, take } from 'rxjs/operators';

import { APP_ENVIRONMENT } from '../../tokens';
import { Environment } from '../../types';
import { ApiError, ApiRequest, ApiRequestActionTypes } from '../actions/api-request.actions';
import { API_TOKEN_SELECTOR } from '../tokens';
import { ApiTokenSelector } from '../types';

export function buildApiHttpRequest<T>(
  baseRequest: HttpRequest<T>,
  env: Environment,
  token: string
) {
  return baseRequest.clone({
    url: `${env.apiUrl}/${baseRequest.url}`,
    setHeaders: { Authorization: `Bearer ${token}` },
  });
}

@Injectable()
export class ApiRequestEffects {
  @Effect()
  processApiRequest$ = this.actions$.pipe(
    ofType<ApiRequest>(ApiRequestActionTypes.ApiRequest),
    flatMap(action => {
      const baseRequest = action.payload.request;
      const actions = action.payload.actions;

      return this.store.pipe(
        select(this.tokenSelector.select),
        take(1),
        flatMap(token => {
          const req = buildApiHttpRequest(baseRequest, this.env, token);

          return concat(
            of(new actions.start()),
            this.http.request(req).pipe(
              filter<any>(res => res instanceof HttpResponse),
              map((res: HttpResponse<any>) => res.body),
              map(data => new actions.success(data)),
              catchError(e => of(new actions.error(e), new ApiError(e)))
            ),
            actions.complete ? of(new actions.complete()) : EMPTY
          );
        })
      );
    }),
    share()
  );

  constructor(
    private actions$: Actions,
    private http: HttpClient,
    private store: Store<any>,
    @Inject(API_TOKEN_SELECTOR) private tokenSelector: ApiTokenSelector,
    @Inject(APP_ENVIRONMENT) private env: Environment
  ) {}
}
