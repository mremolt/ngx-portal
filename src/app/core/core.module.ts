import { CommonModule } from '@angular/common';
import { InjectionToken, NgModule } from '@angular/core';
import { EffectsModule } from '@ngrx/effects';
import { StoreRouterConnectingModule } from '@ngrx/router-store';
import { ActionReducerMap, META_REDUCERS, MetaReducer, StoreModule } from '@ngrx/store';
import { StoreDevtoolsModule } from '@ngrx/store-devtools';

import { environment } from '../../environments/environment';
import { AppEffects } from '../app.effects';
import { RootState, metaReducers, reducers } from '../reducers';

export const REDUCER_TOKEN = new InjectionToken<ActionReducerMap<RootState>>('Registered Reducers');

export function getReducers() {
  return reducers;
}

export function getMetaReducers(): MetaReducer<RootState>[] {
  return metaReducers;
}

@NgModule({
  imports: [
    CommonModule,
    // ngrx
    StoreModule.forRoot(REDUCER_TOKEN),
    !environment.production ? StoreDevtoolsModule.instrument() : [],
    EffectsModule.forRoot([AppEffects]),
    StoreRouterConnectingModule.forRoot(),
  ],
  declarations: [],
  providers: [
    {
      provide: REDUCER_TOKEN,
      useFactory: getReducers,
    },
    {
      provide: META_REDUCERS,
      useFactory: getMetaReducers,
    },
  ],
})
export class CoreModule {}
