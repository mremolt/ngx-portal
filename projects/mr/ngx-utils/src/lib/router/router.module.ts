import { NgModule } from '@angular/core';
import { EffectsModule } from '@ngrx/effects';
import { RouterEffects } from './router.effects';

@NgModule({
  imports: [EffectsModule.forFeature([RouterEffects])],
})
export class RouterModule {}
