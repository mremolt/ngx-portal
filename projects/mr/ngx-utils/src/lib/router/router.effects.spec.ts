import { Location } from '@angular/common';
import { TestBed } from '@angular/core/testing';
import { Router } from '@angular/router';
import { RouterTestingModule } from '@angular/router/testing';
import { provideMockActions } from '@ngrx/effects/testing';
import { ReplaySubject, Subject } from 'rxjs';

import { Back, Forward, Go } from './router.actions';
import { RouterEffects } from './router.effects';

describe('Router Effects', () => {
  let effects: RouterEffects;
  let actions$: Subject<any>;
  let router: Router;
  let loc: Location;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [RouterTestingModule.withRoutes([{ path: 'nowhere', redirectTo: '/' }])],
      providers: [RouterEffects, provideMockActions(() => actions$)],
    });

    effects = TestBed.get(RouterEffects);
    router = TestBed.get(Router);
    loc = TestBed.get(Location);
    actions$ = new ReplaySubject(1);
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('should be created', () => {
    expect(effects).toBeTruthy();
  });

  describe('navigate$', () => {
    it('triggers router navigation', done => {
      const spy = jest.spyOn(router, 'navigate').mockImplementation(() => {});
      actions$.next(new Go({ path: ['nowhere'], query: { id: 42 } }));

      effects.navigate$.subscribe(routing => {
        expect(routing).toMatchSnapshot();
        expect(spy).toHaveBeenCalledWith(['nowhere'], { queryParams: { id: 42 } });
        done();
      });
    });
  });

  describe('navigateBack$', () => {
    it('triggers location change', done => {
      const spy = jest.spyOn(loc, 'back').mockImplementation(() => {});
      actions$.next(new Back());

      effects.navigateBack$.subscribe(() => {
        expect(spy).toHaveBeenCalled();
        done();
      });
    });
  });

  describe('navigateForward$', () => {
    it('triggers location change', done => {
      const spy = jest.spyOn(loc, 'forward').mockImplementation(() => {});
      actions$.next(new Forward());

      effects.navigateForward$.subscribe(() => {
        expect(spy).toHaveBeenCalled();
        done();
      });
    });
  });
});
