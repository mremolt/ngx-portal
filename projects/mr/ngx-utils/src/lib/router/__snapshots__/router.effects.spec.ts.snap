// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Router Effects navigate$ triggers router navigation 1`] = `
Object {
  "path": Array [
    "nowhere",
  ],
  "query": Object {
    "id": 42,
  },
}
`;
